### ISSUE TITLE
Package installation is not working

### ISSUE THREAD

**Issue by Strilanc** [ProjectQ-Framework/ProjectQ#74]
When I run [the install commands given in the documentation](http://projectq.ch/docs/tutorials.html#detailed-instructions-and-os-specific-hints):

    python -m pip install --user projectq

or:

    sudo apt-get install build-essential python3 python3-pip
    sudo pip3 install --user projectq

The install fails with multiple distinct errors:

```
    [...]
        ImportError: No module named pybind11
    [...]
        Command [...]/bin/python -c "import setuptools, tokenize;[...] failed with error code 1 [...]
    [...]
      File "[...]/pip/basecommand.py", line 161, in main
      UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 42: ordinal not in range(128)
```

I think the first two errors are because the documentation says I need to install "build tools" but doesn't actually say what they are or how to install them. So I wasn't able to do that. (I do have `g++` installed.) But the third error looks typical of a python2 data-vs-str encoding issue.

**Comment by Strilanc** [comment]
Interestingly, when I separately run `sudo pip3 install pybind11`, then `sudo pip3 install projectq` works.

Maybe it's not correctly specified as a dependency?

**Comment by Strilanc** [comment]
Repro steps:

```
mkvirtualenv tmp
pip install projectq   #FAILS

pip install pybind11
pip install projectq   #WORKS
```

**Comment by damiansteiger** [comment]
pybind11 is correctly included in the `requirements.txt`:

https://github.com/ProjectQ-Framework/ProjectQ/blob/master/requirements.txt

**Comment by Strilanc** [comment]
Any ideas why pip install projectq would switch from failing to succeeding after installing pybind?

**Comment by thomashaener** [comment]
I have no clue and I think it's really weird that pip does not install pybind despite it being a requirement.

**Comment by Strilanc** [comment]
Matthew thinks it's because of line 30 in setup.py, which has to build an Extension that uses pybind11. But this is before the setup call that actually installs pybind11.

**Comment by maffoo** [comment]
It looks like this could work if we could tell pip to install the required libraries _before_ building the extension module.

**Comment by Strilanc** [comment]
Line 146 creates an exception but doesn't raise it?

**Comment by maffoo** [comment]
I think there might be a problem with the 0.2.0 package that was uploaded to pypi. If I checkout git master and do `pip install .` it works. But `pip install projectq` fails in the way @Strilanc mentioned, and it also fails if I download the 0.2.0 tarfile from pypi and try to run `pip install .` there.

**Comment by maffoo** [comment]
Ok, I think I see the problem. As of [version 6.0.1](https://pip.pypa.io/en/stable/reference/pip_install/#installation-order), `pip` does a topological sort on dependencies before installing them, so that dependencies of package `foo` get installed before `foo` itself. Prior versions of `pip` did not make this guarantee. If I create a new virtualenv on my system (ubuntu 14.04) I get the ancient system version of `pip`, version `1.5.4`. In that case after collecting all the packages, `pip` says:
```
Installing collected packages: projectq, numpy, future, pytest, pybind11, requests, py
  Running setup.py install for projectq
```
So basically it is trying to install `projectq` before `pybind11`, which fails. The solution is to update `pip` itself in the virtualenv:
```
$ pip install --upgrade pip
```
Then when I run `pip install projectq` I get the following:
```
Installing collected packages: numpy, future, py, pytest, pybind11, requests, projectq
```
Crucially, it installs `pybind11` before `projectq` and everything works fine.

**Comment by Strilanc** [comment]
I notice a lot of install instructions often start with `pip install --upgrade pip`, presumably because of stuff like this... makes me think the package should really be able to declare a minimum version of pip.

**Comment by thomashaener** [comment]
Thanks @Strilanc and @maffoo, this is good to know.
Also, I added the missing 'raise' in setup.py. Thanks for pointing this out!

**Comment by maffoo** [comment]
My only other suggestion would be to add a note in the [tutorial](http://projectq.ch/docs/tutorials.html#getting-started) to let users know they might need to update pip before installing projectq itself. Aside from that I think we can close this issue.
